로그인 로직(JwtFilter) 리팩토링

vectorch9 opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
->
## 📌Linked Issues

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## ✏Change Details
- 로그인 로직을 하나의 `JwtFilter`에서 역할을 분리하는 방식으로 변경
> `SpringSecurity`내부의 class, interface를 상속받아 사용하도록 변경
- 현재 불필요하거나 복잡한 코드가 많은데,  정상작동 확인+구조 이해 후 정리할 예정입니다.

<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬Comment
- 크게 세 가지 문제가 있었고 1번째 문제만 해결된 상태입니다.
### 1. 정상적인 JWT 토큰 전달 시 redirect가 무한 반복됨
- 정상적인 JWT를 통해  `JwtAuthenticationFilter`에서 처리후 redirect가 무한반복 되는 현상이 있었음
- 디버깅을 통해 `successfulAuthentication` 메서드 내에서 `super.successfulAuthentication`를 호출 시 '/'로의 redirect가 무한반복되어 문제가 발생함을 인지
- 우선 `super.successfulAuthentication`를 주석처리하여 정상 동작함 -> 이 부분은 super.successfulAuthentication 내부 코드를 확인해본 결과 아예 지워버려도 될 것 같음

### 2. 잘못된 JWT 토큰 전달 시 처리가 안됨
- 잘못된 JWT 토큰 전달/ 빈 JWT 토큰 전달 시 401 오류가 발생 하여야함
- 현재 `SecurityContext`를 clear 시키는건 정상작동
- 그 이후 로그인이 필요한데 Context가 빈 경우 `JwtAuthenticationEntryPoint`에서 에러를 처리해주어야 하지만 처리를 해주지 못하고 있음 -> 결과가 아예 화면에 안뜸
- 디버깅 결과 `JwtAuthenticationEntryPoint`로 아예 코드가 넘어가지 않는 듯함
- **추가로**, `WebSecurityConfig`에서 WHITELIST는 `JwtAuthenticationFilter`를 거칠지 말지, Spring자체의 antMatchers는 로그인이 필요한지 아닌지 구분한뒤 인증이 되지 않았다면 `JwtAuthenticationEntryPoint`로 처리를 넘겨주어야 한다고 이해했는데 맞을까요?

### 3. 한번의 요청시 `JwtAuthenticationFilter`를 두번 이상 거침
- 1번문제와 연관된 문제 같은데, 정상 JWT 토큰 호출 시 `JwtAuthenticationFilter`를 두번 거치게 됩니다. (디버깅 Breakpoint를 걸면  두번씩 걸림을 확인할 수 있습니다) 두번씩 호출될 뿐이지 정상적인 데이터를 return을 해주는데 아직 원인은 모르겠습니다.

vectorch9 committed:
Refactor: 로그인 로직 수정

- SpringSecurity 내부의 class, interface를 상속하여 사용하는 방식으로 변경
- 정상작동하지 않아 추가 수정 필요

vectorch9 committed:
Refactor: 인증 과정 리팩토링

- API단에서 구현되는 인증 과정(ID/PW로그인, 코드를 통한 로그인, 재발급)을 Security filte단에서 작동하도록 수정
- Filter와 Provider간 Token을 통해 정보를 주고 받는 스프링 시큐리티의 기본 설계를 지키도록 리팩토링
- 코드를 통한 로그인 엔드포인트, 일부 response 수정

vectorch9 committed:
Chore: 시큐리티 antMatchers 목록 수정

- 기존엔 각 URL 별로 권한설정을 해줌
- 기본 권한을 ROLE_USER이 필요하도록 설정 해놓고 모든 권한 허용등의 예외 적인 경우만 추가적으로 작성하도록 수정

vectorch9 committed:
Refactor: 로그인한 기기 조회 API에 로그인한 시간 추가

vectorch9 commented:
## ✏Change Details
### 인증 과정 리팩토링
- 인증 과정을 API단 에서 처리하는 것이 아닌 Secuiry Filter단에서 처리하도록 리팩토링
> `SpringSecurity`내부의 Filter, Provider, Token 등을 상속받아 사용하도록 변경
- username/PW를 통한 로그인, 코드를 통한 로그인, 재발급 API를 필터로 옮김
- 추가로, JWT를 통한 인증 과정 필터(기존 JwtFilter)도 Filter, Provider, Token로 나누어 역할을 분담시킴
- filter chain 과정에서 발생하는 `BusinessException`을 처리하기 위해 전용 handle filter도 추가
- 다른 기기에 의해 로그아웃 된 경우 기존에는 ResultCode로 처리하였으나 ErrorCode로 처리하도록 수정
> ResultCode로 처리할 경우 필터에서 successHandler로 두 가지 ResultCode를 넘겨주어야 하는데 추가적인 Handler가 필요해지기 때문에 ErrorCode로 변경하였습니다.

- `SecuriryConfig`의 `antMatchers` 관련 수정 사항
  - 기존에는 각 URI 별로  권한을 설정해주어 가독성이 낮았음
  - 명시되어있지 않은 URI는 "ROLE_USER"이 필요하도록 수정하고, 권한이 필요없거나(로그인 등) 더 높은 권한이 필요한 경우만 직접 명시하도록 수정
> 현잰 swagger, 로그인등 기본적인 경우만 설정해놓았고, 추후에 로그인/비로그인 분리 작업(회원 프로필, 게시물 등) 후 추가해야함
- 인증 과정을 filter로 옮겼지만 Controller에는 로그인 관련 API를 남겨둠(Swagger 상에서 로그인시도를 위해) - 실제 Controller단 까지 요청이 오진 않기 때문에 내부로직은 비어있음
- 로그인 관련 Service 코드는 아직 정리되지 않았는데 후술할 **코드를 통한 비밀번호 재설정**과 코드가 섞여있기 때문에 수정 후 정리할 예정

### 추가 수정 사항
- 로그인한 기기 조회 API에서 마지막으로 로그인한 시간도 반환하도록 수정
- 요청 기기, IP를 조회하는 메서드를 `RequestExtractor` 클래스(Util)로 분리
- 코드를 통한 로그인 API의 엔드포인트를 `/accounts/login/recovery`에서 `/login/recovery`로 수정

## 💬Comment
- 코드를 통한 비밀번호 재설정 문제
   - 해당 API도 filter 단으로 처리하고 싶었는데, filter를 새로 추가하는 것보단 **코드를 통한 로그인 필터**내에서 두 경우로 나누어처리 하는것이 좋을 것같아 해당 부분을 공부 중입니다.
    > 스프링의 `DaoAuthenticationProvider` 내부를 보면 updatePassword 관련 코드가 있는 것으로 보아 함께 처리할 수 있을 것같은데 아직 잘 모르겠네요..
  - 이 부분은 Issue로 남겨두고 좀 더 알아본 후 리팩토링 하겠습니다.
- 금방 끝낼줄 알았는데 시큐리티 자체의 인터페이스, 클래스를 적용하다보니 디버깅하며 시간이 꽤 걸렸네요..

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/MemberAuthController.java:
```diff
@@ -98,51 +98,22 @@ public ResponseEntity<ResultResponse> sendConfirmEmail(
-			return ResponseEntity.ok(ResultResponse.of(REISSUE_SUCCESS, jwtResponse));
-		}
+			return ResponseEntity.ok(ResultResponse.of(REISSUE_SUCCESS, null));
 	}
```
Servlet 파라미터는 빼도 될 것 같습니다!

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/controller/MemberAuthController.java:
```diff
@@ -193,25 +165,12 @@ public ResponseEntity<ResultResponse> resetPassword(@Valid @RequestBody ResetPas
-
-		return ResponseEntity.ok(ResultResponse.of(LOGIN_WITH_CODE_SUCCESS, jwtResponse));
+		return ResponseEntity.ok(ResultResponse.of(LOGIN_WITH_CODE_SUCCESS, null));
 	}
```
여기도 servlet 파라미터는 빼도 될 것 같습니다!

seonpilKim commented on src/main/java/cloneproject/Instagram/global/util/RequestExtractor.java:
```diff
@@ -0,0 +1,26 @@
+
+public class RequestExtractor {
+
+	public static String getDevice(HttpServletRequest request){
+		return request.getHeader("user-agent");
+	}
+
+	public static String getClientIP(HttpServletRequest request) {
+		String ip = request.getHeader("X-Forwarded-For");
+		if (ip == null)
+			ip = request.getHeader("Proxy-Client-IP");
+		if (ip == null)
+			ip = request.getHeader("WL-Proxy-Client-IP");
+		if (ip == null)
+			ip = request.getHeader("HTTP_CLIENT_IP");
+		if (ip == null)
+			ip = request.getHeader("HTTP_X_FORWARDED_FOR");
+		if (ip == null)
+			ip = request.getRemoteAddr();
+		return ip;
+	}
```
컨벤션 적용이 안 되어 있네요. 그리고 MemberAuthController에 중복 코드가 있는 것 같아요.

vectorch9 replied:
작동 확인후 빼려했는데 까먹었네요..

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/member/service/ResetPasswordCodeUserDetailService.java:
```diff
@@ -0,0 +1,49 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ResetPasswordCodeUserDetailService implements UserDetailsService {
+
+	private final ResetPasswordCodeRedisRepository resetPasswordCodeRedisRepository;
+	private final MemberRepository memberRepository;
+
+	@Override
+	@Transactional(readOnly = true)
+	public UserDetails loadUserByUsername(String username) {
+		return resetPasswordCodeRedisRepository.findByUsername(username)
+			.map(this::createUserDetails)
+			.orElseThrow(PasswordResetFailException::new);
+	}
+
+	private UserDetails createUserDetails(ResetPasswordCode resetPasswordCode) {
+		Member member = memberRepository.findByUsername(resetPasswordCode.getUsername())
+			.orElseThrow(() -> new EntityNotFoundException(ErrorCode.MEMBER_NOT_FOUND));
+		final GrantedAuthority grantedAuthority = new SimpleGrantedAuthority(member.getRole().toString());
+		return new User(
+			String.valueOf(member.getId()),
+			resetPasswordCode.getCode(),
+			Collections.singleton(grantedAuthority));
+	}
+
+}
```
혹시 이 클래스가 사용되는 곳이 있나요?/어디에 사용될 예정인가요?
그리고 JwtUserDetailsService 클래스도 사용이 되지 않는데 어떤 이유로 구현하신 건지 궁금합니다!

vectorch9 replied:
- 각각 `DaoAuthenticationProvider`, `ResetPasswordCodeAuthenticationProvider`에서 사용됩니다. `DaoAuthenticationProvider`는 스프링 자체에서 제공하는 Provider인데 이름대로 Dao(DB)의 정보를 제공받아 인증과정을 거치게 됩니다. 이 과정에서 `UserDetailsService`의 구현체가 필요해 `JwtUserDetailsService`를 주입합니다.
- `ResetPasswordCodeAuthenticationProvider`의 경우도 `DaoAuthenticationProvider`와 비슷한 로직을 갖기때문에 동일하게 `UserDetailsService`를 주입받도록 설계하였습니다. `WebSecurityConfig`에 Bean 등록하는 코드부분을 찾아보시면 두 서비스를 주입해주고 있습니다!
> 추가로 `JwtUserDetailsService` 네이밍은 처음 로그인 구현할 때 작성했는데, 사실 JWT와는 관련이 없으므로 `CustomUserDetailsService`로 수정하겠습니다

seonpilKim replied:
아아 넵 이해했습니다. 감사합니다 :)

추가로 CustomUsernamePasswordAuthenticationFilter는 어떤 이유로 추가하신 건지 궁금합니다!

vectorch9 replied:
처음엔 시큐리티의 `UsernamePasswordAuthenticationFilter`를 그대로 사용했습니다. 해당 메서드를 보시면 아시겠지만 request에서 `getParameter`를 통해 username, password를 받아오게 됩니다. 로그인의 경우 비밀번호가 노출되면 안되므로 body를 통해서 전달해야하는데 request의 getParamter 함수는 body에서 정보를 받을 수 없습니다. 따라서, ObjectMapper를 통해 body를 LoginRequest(객체)로 변환 시킨 후 username, password를 받기 위해 `CustomUsernamePasswordAuthenticationFilter`를 따로 만들었습니다.

seonpilKim replied:
그럼 기존 API와 동일하게 Json으로 username, password를 받아서 처리하려는 목적으로 커스터마이징 하신 걸로 이해하면 될까요?

vectorch9 replied:
넵 맞습니다 !

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/WebSecurityConfig.java:
```diff
@@ -1,108 +1,215 @@
+		map.put("/reissue", ResultCode.REISSUE_SUCCESS);
+		map.put("/login/recovery", ResultCode.LOGIN_WITH_CODE_SUCCESS);
+		customAuthenticationSuccessHandler.setResultCodeMap(map);
+	}
```
```diff
- Map<String, ResultCode> map = new HashMap<>();
+ final Map<String, ResultCode> map = new HashMap<>();
```
final 키워드 추가해 주면 좋을 것 같아요~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/WebSecurityConfig.java:
```diff
@@ -1,108 +1,215 @@
+	}
+
+	@Bean
+	public CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter() throws Exception {
+		final CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter = new CustomUsernamePasswordAuthenticationFilter();
+		customUsernamePasswordAuthenticationFilter.setAuthenticationManager(super.authenticationManager());
+		customUsernamePasswordAuthenticationFilter.setAuthenticationSuccessHandler(customAuthenticationSuccessHandler);
+		customUsernamePasswordAuthenticationFilter.setAuthenticationFailureHandler(customAuthenticationFailureHandler);
+		return customUsernamePasswordAuthenticationFilter;
+	}
```
혹시 여기 한 줄에 120글자 초과 안 하던가요?

vectorch9 replied:
확인해보니까 딱 118글자네요. 그래도 가독성 떨어지니까 두줄로 만들겠습니다

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/provider/ResetPasswordCodeAuthenticationProvider.java:
```diff
@@ -0,0 +1,85 @@
+
+	private final EmailCodeService emailCodeService;
+
+	protected void additionalAuthenticationChecks(
+		UserDetails userDetails, UsernamePasswordAuthenticationToken authentication) throws AuthenticationException {
+		if (authentication.getCredentials() == null) {
+			this.logger.debug("Failed to authenticate since no credentials provided");
+			throw new PasswordResetFailException();
+		} else {
+			String presentedEmailCode = authentication.getCredentials().toString();
+			if (!presentedEmailCode.equals(userDetails.getPassword())) {
+				this.logger.debug("Failed to authenticate since email code does not match stored value");
+				throw new PasswordResetFailException();
+			}
+		}
+	}
+
+	protected void doAfterPropertiesSet() {
+		Assert.notNull(this.userDetailsService, "A UserDetailsService must be set");
+	}
+
+	protected final UserDetails retrieveUser(String username, UsernamePasswordAuthenticationToken authentication) throws
+		AuthenticationException {
+		try {
+			UserDetails loadedUser = this.getUserDetailsService().loadUserByUsername(username);
+			if (loadedUser == null) {
+				throw new EntityNotFoundException(MEMBER_NOT_FOUND);
+			} else {
+				return loadedUser;
+			}
+		} catch (InternalAuthenticationServiceException e) {
+			throw e;
+		}
+	}
```
final 키워드 추가하면 좋을 것 같습니다~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/provider/ResetPasswordCodeAuthenticationProvider.java:
```diff
@@ -0,0 +1,85 @@
+@RequiredArgsConstructor
+public class ResetPasswordCodeAuthenticationProvider extends AbstractUserDetailsAuthenticationProvider {
+
+	private static final String USER_NOT_FOUND_PASSWORD = "userNotFoundPassword";
```
이건 어떤 이유로 추가하신 걸까요?

vectorch9 replied:
시큐리티 `DaoAuthenticationProvider`에서 코드를 그대로 긁어와서 커스텀했는데 빼는걸 빠트렸네요! 빼겠습니다

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/provider/ResetPasswordCodeAuthenticationProvider.java:
```diff
@@ -0,0 +1,85 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public class ResetPasswordCodeAuthenticationProvider extends AbstractUserDetailsAuthenticationProvider {
```
전체적으로 오버라이딩 하는 메소드는 `@Override` 추가하면 좋을 것 같아요~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/provider/ReissueAuthenticationProvider.java:
```diff
@@ -0,0 +1,50 @@
+		return authenticated;
+	}
+
+	protected void deleteRefreshToken(RefreshToken refreshToken) {
```
혹시 protected로 제한한 이유가 있을까요?

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/ResetPasswordCodeAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+public class ResetPasswordCodeAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
+	private static final AntPathRequestMatcher ANT_PATH_REQUEST_MATCHER = new AntPathRequestMatcher("/login/recovery",
+		"POST");
+	ObjectMapper objectMapper = new ObjectMapper();
```
final~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/ResetPasswordCodeAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+import cloneproject.Instagram.global.config.security.token.ResetPasswordCodeAuthenticationToken;
+
+public class ResetPasswordCodeAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
+	private static final AntPathRequestMatcher ANT_PATH_REQUEST_MATCHER = new AntPathRequestMatcher("/login/recovery",
```
한 줄 띄우기~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/ResetPasswordCodeAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+import cloneproject.Instagram.domain.member.dto.LoginWithCodeRequest;
+import cloneproject.Instagram.global.config.security.token.ResetPasswordCodeAuthenticationToken;
+
+public class ResetPasswordCodeAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
```
오버라이딩한 메소드 `@Override` 추가해 주세요~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/ReissueAuthenticationFilter.java:
```diff
@@ -0,0 +1,40 @@
+import cloneproject.Instagram.domain.member.exception.JwtInvalidException;
+import cloneproject.Instagram.global.config.security.token.ReissueAuthenticationToken;
+
+public class ReissueAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
```
한 줄 띄우기, `@Override` 추가~

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,56 @@
+	protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
+		AuthenticationException failed) throws IOException, ServletException {
+		SecurityContextHolder.clearContext();
+		super.unsuccessfulAuthentication(request, response, failed);
```
중복코드👀

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,56 @@
+	public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws
+		AuthenticationException {
+		final String authorizationHeader = request.getHeader(AUTHORIZATION_HEADER);
+		final String jwt;
+		jwt = jwtUtil.extractJwt(authorizationHeader);
```
코드 합치기👀

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/CustomUsernamePasswordAuthenticationFilter.java:
```diff
@@ -0,0 +1,46 @@
+
+import cloneproject.Instagram.domain.member.dto.LoginRequest;
+
+public class CustomUsernamePasswordAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
```
한 줄 띄우기, final 추가, `@Override` 추가👀

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/WebSecurityConfig.java:
```diff
@@ -1,108 +1,215 @@
-                .antMatchers("/chat/rooms/**").hasAuthority("ROLE_USER");
-
-    }
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
```
한 줄 띄우기👀

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/WebSecurityConfig.java:
```diff
@@ -1,108 +1,215 @@
+
+	@Bean
+	public CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter() throws Exception {
+		final CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter = new CustomUsernamePasswordAuthenticationFilter();
```
간단히 변수 네이밍을 filter로 사용하면 더 보기 편할 것 같아요! (아래 메소드도 동일)

seonpilKim commented on src/main/java/cloneproject/Instagram/global/config/security/filter/CustomExceptionHandleFilter.java:
```diff
@@ -0,0 +1,40 @@
+import cloneproject.Instagram.global.error.exception.BusinessException;
+
+@Component
+public class CustomExceptionHandleFilter extends OncePerRequestFilter {
```
이 필터는 ResetPasswordCodeAuthenticationFilter, ReissueAuthenticationFilter, CustomUsernamePasswordAuthenticationFilter 에서 던지는 예외를 처리하기 위해 추가한 필터로 이해하면 될까요?

vectorch9 replied:
넵 맞습니다. 시큐리티 자체에서 쓸수 있는 `AuthenticationFailureHandler`는 `AuthenticationException`만 잡아주기 때문에 BusinessException을 처리하기 힘듭니다. 일부 BusinessException을 AuthenticationException로 만들거나 내부에 포함시키는 시도를 해봤는데 코드 가독성이 떨어지고 예외처리 시 놓치는 부분이 많이 생겨 하나의 filter를 추가로 넣는게 제일 좋다고 판단했습니다.

seonpilKim reviewed (commented):
코드 보면서 많이 고민해 보시고 짜신 것 같다는 느낌이 들었어요👍
전체적으로 컨벤션 점검도 부탁드립니다!

vectorch9 committed:
Refactor: MemberAuthController 컨벤션 적용, 불필요 코드 삭제
- MemberAuthController, RequestExtractor에 컨벤션 적용
- 불필요 메서드, 매개변수 삭제
- MemberAuthController내 일부 API는 실제론 Controller가 아닌 filter에서 처리해야 하기 때문에 전용 Exception을 추가

vectorch9 committed:
Style: 인증 관련 코드에 컨벤션 전체적으로 적용
- 컨벤션 적용
- 컨벤션에는 맞지만 일부 가독성 떨어지는 코드 수정 (한줄에 120글자 제한이지만 115 글자 정도만 넘어도 가독성이 떨어져 수정)

vectorch9 commented:
말씀주신부분 수정했습니다!
Intellij에서 지원하는 "다른 코드에서 1회 사용"으로 반줄씩 띄워져 있어서 빈줄이 있는줄 알았는데 다 없었네요...

vectorch9 committed:
Style: JwtUserDetailsService의 네이밍 변경
- JwtUserDetailsService -> CustomUserDetailsService로 네이밍 변경
